### ISSUE TITLE
Enhancing node representations

### ISSUE THREAD

**Issue by bigaidream** [rahul99/msc_plan#11]
https://arxiv.org/abs/1609.05600

**Comment by rahul99** [comment]
This paper seems like a different project (details below) but uses pre-trained word vectors from Stanford that we can leverage:

This paper takes as input a scene image and a question in text form and encodes both as a graph representation as follows:
a. For questions, each word acts as one node and edges as relation between different words. This info is derived from Stanford pretrained vocab of 2Billion words (https://nlp.stanford.edu/projects/glove/)
b. Every object (including pose of the person) is represented as one-hot vector is already assumed to be available.

This work then learns a GRU recurrent model independently over scene description and the question graph for the purpose of VQA (so the objective is slightly different from what we hope to achieve)

Usage in GAN:
Instead of constructing the scene-triplet into adjacency matrix, which is very sparse, we can use the pretrain-word vocabulary to represent scene information into a graph embedding. This together with the input image will be used to train conditional three player game (similar to ALI: https://arxiv.org/pdf/1606.00704.pdf) as follows:
a. Forward model/Decoder (noise + graph embedding*) -----> Generated Image
b. Reverse Model/Encoder (original image) ------> jointly learn noise and graph embedding
c. Discriminator ----> Trained to maximize likelyhood of reverse model while generator is trained to 
maximize likelyhood of the forward model.

After training is complete, we discard the Discriminator and the forward model and use Decoder for future image generation.

[Proposed change:] We have now two ways to go about using the graph information:
1. Use the pretrained stanford word vocab to generate word representation. Learn 1D embedding to encode this as opposed to 2D embedding (of the same shape as inp image)

2. We can use GRU recurrent models, similar to this paper to encode graph information.

For both these embeddings, since these are linear, we add this to the intermediate flattened layer of the discriminator for prediction. 

[Current progress]:
I am looking into step one. Still have not fetched word representation from pretrained stanford vocab. After this step, we can use one of the standard 1D embedding schemes mentioned here as I shared with you previously (https://iapr-tc15.greyc.fr/download/03MT.pdf)

**Comment by rahul99** [comment]
On embedding, there are two issues that I encountered today:
1. Some of the words were missing in the pre-trained vocab, for example, "On" 
     Workaround: I can train model (Stanford Glove) again on the graph data for VG. Tried for some sample graph data files.

2. Once every word is represented as a vector of some dim (300 default). How to find a vector embedding per image which has many triplets, where each triplet has minimum three words?
     [In VQ paper]: They directly train an LSTM on this graph data. 

     [Our case]: For vector embedding, we can approach as follows:
            a. Collect the text descriptions for a given image (done before)
            b. Train the (Stanford) Glove on these descriptions to have vector embedding for every word 
            c. Collect all the triplets for a given image (VG scene description and sentence parser we have 
                customized before)
            d. For each entity in the triplet, get the word(s) embedding and concatenate all of it together to 
                get a single vector representative of that triplet.
            e. Now, some triplets may be trivial. So we need to get rid of them. In effect, in the collection of these new "large" vectors, we need to discard less important ones. Here we can fit an SVD or any other upsupervised feature reduction/selection techinque to retain top-1 (or concat of top-2) vector is the final graph embedding for that image.

Let me know if step [e.], looks reasonable as I thought of it myself (I could not find any reference that can handle this kind of multi-matrix representation)

**Comment by bigaidream** [comment]
[Visual Word2Vec](https://github.com/satwikkottur/VisualWord2Vec)

[Visual Relationship Detection](https://github.com/Prof-Lu-Cewu/Visual-Relationship-Detection)

https://arxiv.org/abs/1707.05427
